### PR TITLE
JP-3810: Add tests, fix bugs, remove unused code for AMI module

### LINKED ISSUE #8994: Add more unit tests for AMI steps
_Issue [JP-3810](https://jira.stsci.edu/browse/JP-3810) was created on JIRA by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke):_

Unit test coverage for AMI steps in calwebb_ami3 is a little light: we could use some more tests to make sure we catch all corner cases and possible regresssions.

### PR DESCRIPTION
**Pull request by emolter**
Resolves [JP-3810](https://jira.stsci.edu/browse/JP-3810)

Closes #8994 

This PR adds a unit testing suite to the AMI module.  In so doing, lots of unused code was found, as well as several bugs which have been fixed in this PR.  I have tried to label the major changes with in-line comments so my thought process is clear and to promote discussions on those specific changes.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

File: jwst/ami/nrm_core.py [hunk 1]
```diff
@@ -165,6 +153,10 @@ def fit_fringes_single_integration(self, slc):
         """
         Generate the best model to match a single slice.
 
+        TODO: Is it a bug that the LgModel is being initialized with the
+        zeroth wavelength every time?  When we iterate over this, it's an iteration over
+        wavelength.
```
Comment (emolter) [inline jwst/ami/nrm_core.py:158]: @rcooper295 Can you comment on whether it is a bug or not that when a single slice is fitted, it's always using `bandpass=self.instrument_data.wls[0]`, but [this line](https://github.com/spacetelescope/jwst/blob/7a3fc5f3667eec4dffe6a4ba7cd76fd9e1c28122/jwst/ami/nrm_core.py#L101) in `fit_fringes_all` appears to call this in a `for` loop iterating over wavelength?

File: jwst/ami/tests/test_ami_analyze.py
Comment (emolter) [inline jwst/ami/tests/test_ami_analyze.py:1]: I reorganized these tests to have one file per module

File: jwst/ami/utils.py
Comment (emolter) [inline jwst/ami/utils.py:1]: Mostly just removing unused functions here.

I retained all methods of the `Affine2d` class even though several are unused.  I would be very happy to be told that these could be removed though :)

File: jwst/ami/matrix_dft.py [hunk 2]
```diff
@@ -35,18 +34,13 @@
 
 Examples
 --------
-mf = matrixDFT.MatrixFourierTransform()
-result = mf.perform(pupilArray, focalplane_size, focalplane_npix)
+result = matrix_dft.matrix_dft(pupilArray, focalplane_size, focalplane_npix)
 """
 
-__all__ = ["MatrixFourierTransform"]
+__all__ = ["matrix_dft", "matrix_idft"]
```
Comment (emolter) [inline jwst/ami/matrix_dft.py:40]: Retained the inverse function even though it's unused, both for ease of testing and because it's relatively simply implemented

File: jwst/ami/utils.py [hunk 3]
```diff
@@ -323,13 +282,9 @@ def makedisk(n, r, ctr=(0, 0)):
     ----------
     n : int
         Size of 1 dimension of the array to be returned
-
     r : int
         Radius of disk
 
-    ctr : (int, int)
```
Comment (emolter) [inline jwst/ami/utils.py:330]: centering did not work, and since it was unused I removed it

File: jwst/ami/utils.py [hunk 4]
```diff
@@ -467,23 +356,14 @@ def min_distance_to_edge(img, cntrimg=False):
...
     -------
     peakx, peaky : integer, integer
         Coordinates of the peak pixel
-
     h : integer
         Distance to the nearest image edge
     """
-    if cntrimg is True:
-        # Only look for the peak pixel at the center of the image
-        ann = makedisk(img.shape[0], 31)  # search radius around array center
```
Comment (emolter) [inline jwst/ami/utils.py:483]: ctrimg=True was unused so I removed it.  It made me uncomfortable that this was hard-coded anyway

File: jwst/ami/utils.py [hunk 5]
```diff
@@ -529,54 +409,46 @@ def find_centroid(a):
...
+        Extremum of the quadratic
     """
-    y_max = -p[1] / (2.0 * p[0]), -p[1] * p[1] / (4.0 * p[0]) + p[2]
-
-    return y_max
+    return -p[1] / (2.0 * p[0]), -p[1] * p[1] / (4.0 * p[0]) + p[2]
 
 
-def findpeak_1d(yvec, xvec):
+def findpeak_1d(xvec, yvec):
```
Comment (emolter) [inline jwst/ami/utils.py:442]: These inputs were flipped here, but also flipped where this was called, so I flipped both to make this self-consistent and easier to test

File: jwst/ami/utils.py [hunk 6]
```diff
@@ -1296,39 +922,39 @@ def handle_bandpass(bandpass, throughput_model):
     Parameters
     ----------
     bandpass : Synphot spectrum or array, or None
```
Comment (emolter) [inline jwst/ami/utils.py:924]: Fixed a bug in this function that caused output to have the wrong units if the input was a synphot spectrum

File: jwst/ami/ami_average_step.py [hunk 7]
```diff
@@ -4,6 +4,10 @@
 __all__ = ["AmiAverageStep"]
 
 
+# TODO: What is the history of this step? It's not used in the pipeline.
+# Was it superseded or does it remain useful?
```
[the remark this answers is not in the document]
Comment (rcooper295) [reply jwst/ami/ami_average_step.py:8]: We used to average together the products from multiple exposures (e.g. dither patterns), but then realized we weren't sure if that would actually improve the quality of the interferometric observables we're extracting or make it more difficult to calibrate out instrumental effects, so we deprecated it for now but left the structure in place so that if we wanted to bring it back we wouldn't have to start from scratch. There aren't currently any plans to revive it as far as I'm aware.

[the remark this answers is not in the document]
Comment (emolter) [reply jwst/ami/ami_average_step.py:8]: Good to know.  Because GitHub stores old versions of code, this would still be accessible in the future even if we did delete it.  Would you be ok with adding a DeprecationWarning to the step in this build, then removing it in the next build?

Just for my own understanding, does the averaging in the AmiAverageStep differ from what's done by [oifits.RawOifits.average_observables()](https://github.com/spacetelescope/jwst/blob/e12dfde97ed99be0e549efeafca4e54142222dab/jwst/ami/oifits.py#L117)?

File: jwst/ami/leastsqnrm.py [hunk 8]
```diff
@@ -69,12 +69,12 @@ def weighted_operations(img, model, dqm=None):
...
     # Remove not-to-be-fit data from the flattened "img" data vector
     flatimg = img.reshape(np.shape(img)[0] * np.shape(img)[1])
-    flatdqm = dqm.reshape(np.shape(img)[0] * np.shape(img)[1])
 
     if dqm is not None:
+        flatdqm = dqm.reshape(np.shape(img)[0] * np.shape(img)[1])
         nanlist = np.where(flatdqm)  # where DO_NOT_USE up.
     else:
-        nanlist = (np.array(()),)  # shouldn't occur w/MAST JWST data
+        nanlist = (np.array((), dtype=np.int32),)  # shouldn't occur w/MAST JWST data
```
Comment (emolter) [inline jwst/ami/leastsqnrm.py:77]: These changes made it so that `dqm=None` was a valid input to the function, which did not work before

**Comment by emolter** [comment]
initial round of regression tests https://github.com/spacetelescope/RegressionTests/actions/runs/14086839639

**Comment by emolter** [comment]
There are still a few small outstanding things I need to run past the INS side, but I'm going to go ahead and mark this ready for review.  It would be helpful if someone from Chartreuse could look through this from a code perspective.

File: jwst/ami/ami_average_step.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rcooper295) [reply jwst/ami/ami_average_step.py:9]: I think that would be fine; if we need to we can dig it up from old versions. 
There is a slight difference. The AmiAverageStep was for averaging across exposures (files), which could be from different positions on the detector, while the average_observables() function you pointed out is for averaging different integrations within an exposure (file). Since taking a series of e.g. 5 exposures of one integration each at the same position isn't functionally really any different from taking a single exposure containing 5 integrations (except for overheads), there is some overlap.

File: jwst/ami/bp_fix.py [hunk 9]
```diff
@@ -4,7 +4,7 @@
 import logging
 
 from copy import deepcopy
-from poppy import matrixDFT
+from .matrix_dft import matrix_dft
```
[the remark this answers is not in the document]
Comment (rcooper295) [reply jwst/ami/bp_fix.py:7]: I think these two versions have the opposite sign convention -- see the sign on the imaginary numbers in [these lines](https://github.com/spacetelescope/poppy/blob/develop/poppy/matrixDFT.py#L212-L218) in the poppy version compared to the equivalent lines in the AMI matrix_dft. I've asked our lead to weigh in on that as well.

[the remark this answers is not in the document]
Comment (emolter) [reply jwst/ami/bp_fix.py:7]: Good to know.  If that is the case, is there any way to make a change such that we only need one version of `matrix_dft`?  Right now, both are called in different places.  Perhaps this is achievable either by adding an option to toggle the sign, or even to modify the input/output to flip sign before/after the call?

If this is a lot of work it's not a big deal, but from a software maintenance perspective it's really nice both to avoid having an imported function that's very similar to an in-house function, and to be able to remove a dependency.

[the remark this answers is not in the document]
Comment (tapastro) [reply jwst/ami/bp_fix.py:7]: Big 👍 on any possibility of dropping `poppy` as a dependency! It pulls in a few other packages to our dependency list.

**Comment by anand0xff** [comment]
I used the -1j forward transform convention in the routine webbpsf is built
around - matrixDFT. Years later the JWST project decided to unify on the
other one. So Marshall changed poppy to switch the sign in the forward
transform.  It makes a difference in the way tilt works.
Only one convention should be used. I suggest keeping matrixDFT local to
AMI code if we want less dependence on other modules.
Any sign you decide on is fine - maybe the doc string can indicate that a
negative slope for eg phi(x) will move the PSF to the X. X is either left
or right in a DS9 view of the PSF. If you want to use an imshow left or
right whichever is correct) please do so, and state that in the appropriate
docstring.
I’m happy see you going through this in detail!

Ned, fyi the code should add a column slope in the model fitting to deal
with 1/f noise.  That has not made it to the Institute high priority to do
list. If you add those parameters to the fit let us know. If I do it I’ll
let you know. That adds as many slope parameters as columns used in the
fit.

Thanks!

Anand

**Comment by drlaw1558** [comment]
@anand0xff Following up quickly on your comment about adding a column slope to deal with 1/f noise; I don't think I recall hearing about this?  To what extent is the clean_flicker_noise step useful for AMI observations?  Are there any limitations on its ability to correct 1/f noise for this mode?

**Comment by anand0xff** [comment]
The image from Ami fills the subarray. That makes things more complicated
for a column slope. We are not spending time on this for AMI now as we have
other things that take priority.

**Comment by emolter** [comment]
@anand0xff Thanks for the helpful response.  I have a few questions to make sure I understand this fully.

There are three uses of `matrix_dft` in the pipeline, as follows:
- Poppy version: `bp_fix.transform_image`, used in `calc_pupil_support`
- Poppy version: `bp_fix.calcpsf`, used in `calc_pupil_support`
- Jwst version: `utils.find_centroid`, which itself is only used in `FringeFitter` if `psf_offset_ff` is not set.

It looks to me like `calc_pupil_support`, which itself is used in `fix_bad_pixels`, Fourier transforms the pupil mask into a PSF, then Fourier transforms it again to be used for the bad pixel identification.  Are these two uses directly inverting one another?  (If so, it looks like we need not worry about this causing any unintended consequences).

The use in `utils.find_centroid` looks like it's the one you're talking about where I can check the sign of the tilts; is that right?  It looks like there is already a note in there about the expected convention; it says
```
    XY conventions meshed to lg_model conventions:
    if you simulate a psf with pixel_offset = ( (0.2, 0.4), ) then blind
        application  centroid = utils.find_centroid()

    Returns the image centroid (0.40036, 0.2000093) pixels in image space.
```
I will check if this is still correct, and if so, I'll see if it makes sense to also update the docstring of `matrix_dft` itself.

File: jwst/ami/lg_model.py [hunk 10]
```diff
@@ -308,60 +241,28 @@ def fit_image(
         reference image (a cropped deNaNed version of the data) to run
         correlations. It is recommended that the symmetric part of the data be
         used to avoid piston confusion in scaling.
+        TODO: change name of self.singvals or self.linfit_results to be the same, for consistency.
+        This would be easier if matrix_operations and weighted_operations both did their fitting
+        with scipy
+        TODO: weighted mode is never used by the pipeline. is that still needed?
```
[the remark this answers is not in the document]
Comment (rcooper295) [reply jwst/ami/lg_model.py:247]: I think we do want to keep this option; we played around with it in the standalone version of this code early on and got better results when using the unweighted (matrix operations) verision, but we may want to go back to using it at some point in the future, and it's good to have the option for testing.

File: jwst/ami/lg_model.py [hunk 11]
```diff
@@ -308,60 +241,28 @@ def fit_image(
...
             Bad pixel mask of same dimensions as image
-
         weighted : bool
             Use weighted operations in the least squares routine
         """
-        self.model_in = model_in
         self.weighted = weighted
-        self.saveval = savepsfs
-
-        if model_in is None:  # No model provided
```
[the remark this answers is not in the document]
Comment (rcooper295) [reply jwst/ami/lg_model.py:336]: I'm not sure if there was a reason for this kind of circular input/output originally, but if that's not used by the pipeline it is probably safe to make this change.

File: jwst/ami/matrix_dft.py (hunk 2 above)
Comment (rcooper295) [reply jwst/ami/matrix_dft.py:40]: Thanks, I think it's good to have both the forward and reverse transforms!

Comment (anand0xff) [reply jwst/ami/matrix_dft.py:40]: I agree its good to leave in, as discussed above

**Comment by rcooper295** [comment]
Thanks for checking on those usages @emolter and for the confirmation @anand0xff! Overall, the rest of the changes look like they will greatly improve the readability of the code. There are a few options we toggle that I think we want to keep (e.g. weighted fit) even though they aren't currently accessible to users from the step level. I've asked for confirmation that the rest of the currently unused functions are ok to remove as well, or if they're useful to leave for any testing workflows.

File: jwst/ami/bp_fix.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (emolter) [reply jwst/ami/bp_fix.py:7]: Setting this to Resolved based on Anand's comments below, which state that the sign convention does not matter in `bp_fix` and this change should be safe.

**Comment by emolter** [comment]
Thanks for the info @anand0xff, your explanation makes total sense and answers the question of why changing this didn't cause any of my tests to change behavior/fail.  It sounds like we can safely remove the `poppy` dependency, which is great!

@rcooper295 Thanks so much for your help with this, and for coordinating these changes with your team.  I'll retain the weighted fit option and any others you specify in the other comments once you've talked to the rest of the team.

Hopefully all this work will make future performance improvements or additional features (e.g. 1/f noise fitting) much easier to implement.

**Comment by rcooper295** [comment]
Hi @emolter, sorry for the delay -- we're fine with removing all the remaining unused functions; we can always add them to external testing tools or revive them in the code here if needed.

File: jwst/ami/leastsqnrm.py [hunk 12]
```diff
@@ -595,39 +219,33 @@ def matrix_operations(img, model, flux=None, linfit=False, dqm=None):
     log.info("flat img * transpose dimensions %s", np.shape(inverse))
 
     if linfit:
-        try:
-            from linearfit import linearfit  # type: ignore[import-not-found]
```
[the remark this answers is not in the document]
Comment (rcooper295) [reply jwst/ami/leastsqnrm.py:599]: Since it was easy to port over (and you have already done so) let's leave this as an option for now.

File: jwst/ami/hextransformee.py [hunk 13]
```diff
@@ -69,41 +58,40 @@ def gfunction(xi, eta, **kwargs):
...
 
 
-def hextransform(s=None, c=None, d=None, lam=None, pitch=None, affine2d=None):
+def hextransform(s, d, lam, pitch, affine2d, c=None):
     """
     Calculate the complex array analytical transform of a (distorted if necessary) hexagon.
 
+    TODO: Why does this function force the center pixel to have value sqrt(3)/2
+    but does not normalize the rest of the PSF accordingly? Is this intended behavior
+    or is it a bug?
```
[the remark this answers is not in the document]
Comment (emolter) [reply jwst/ami/hextransformee.py:67]: @rcooper295 @anand0xff I am still wondering about this.  If nobody knows, we can just leave this comment as a TODO when the branch gets merged - we wouldn't want to let this minor thing get in the way of the rest of the PR

File: jwst/ami/nrm_core.py (hunk 1 above)
Comment (emolter) [reply jwst/ami/nrm_core.py]: The above comment is not correct - no iteration over wavelength was being done.  The accessing of `instrument.data.nwav` in the linked line was simply being used to get the number of slices, since `nwav == nslices`.  This was changed to use `nslices` to clarify the intent of the loop, which is to iterate over slices.

**Comment by tapastro** [comment]
Code updates look good, though the PR has accumulated some unit test failures that should be addressed. Sorry to have left it sitting so long!

**Comment by emolter** [comment]
> the PR has accumulated some unit test failures that should be addressed. Sorry to have left it sitting so long!

No worries!  They would have started failing eventually anyway

File: jwst/ami/tests/test_matrix_dft.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    assert fraction_close > 0.99
+
+    # test nonzero offset
+    offset = (10, -7)
+    dft_offset = matrix_dft.matrix_dft(circular_pupil, nlam_d, npix, centering="ADJUSTABLE", offset=offset)
+    # ensure max value of dft is at the offset from center of array
+    argmax_2d = np.unravel_index(np.argmax(np.abs(dft_offset)), dft.shape)
+    center = np.array(dft_offset.shape)/2
+    offset_idx = center + np.array(offset) - 1
+    assert(all(argmax_2d == offset_idx))
```
Comment (melanieclarke) [inline jwst/ami/tests/test_matrix_dft.py:32]: @emolter - after merging this to main, this test is failing on my Mac with numpy>2, passing with numpy<2.  The second value appears to be off by 1 with numpy 2.  Did you see this in testing?

Comment (rcooper295) [reply jwst/ami/tests/test_matrix_dft.py:32]: We'll take a closer look at this on the AMI team as well this week to see if we can figure out what is going on.

File: jwst/ami/hextransformee.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (rcooper295) [reply jwst/ami/hextransformee.py:67]: I asked Anand about this -- the analytical calculation of the hexagon transform has a 1/x numerical singularity at the center and at two other symmetrical lines, so we just replace it with what we know to be the pixel value at that location, so this is the intended behavior.

[the remark this answers is not in the document]
Comment (emolter) [reply jwst/ami/hextransformee.py:67]: Great thanks for following up.  This TODO can be removed then, perhaps as part of the PR that fixes the unit test we are discussing.  I'm happy to do it, or you can

File: jwst/ami/tests/test_matrix_dft.py (hunk 14 above)
Comment (emolter) [reply jwst/ami/tests/test_matrix_dft.py:32]: I figured out the problem- see https://github.com/spacetelescope/jwst/issues/9437 
There's nothing wrong with the matrix DFT code, the test was just badly written.  My fault.  Will submit a follow-up PR.

Comment (rcooper295) [reply jwst/ami/tests/test_matrix_dft.py:32]: Thanks Ned! Anand and I did look at this together today and concluded that since I think we only ever use the "ADJUSTABLE" default (with no offset), it might make sense to remove the other centering options from this implementation of matrix_dft and then we wouldn't need those other parts of the test. I'll leave it up to you whether you want to make those extra changes now that you've solved the issue anyways; functionally I don't think it should make a difference to us.

Comment (emolter) [reply jwst/ami/tests/test_matrix_dft.py:32]: @rcooper295 To clarify, does that mean you would be ok with (and maybe even prefer?) removing both `offset` and `centering` as options, and hard-coding the behavior to mimic the current behavior with `centering="ADJUSTABLE"` and `offset=(0,0)`?

Comment (rcooper295) [reply jwst/ami/tests/test_matrix_dft.py:32]: Yes, that's what I was suggesting. From a coding standpoint it seems like it would be cleaner to keep only the options that we actually use, but from the perspective of (future) flexibility and covering all our bases maybe we should just leave it now that you solved the issue with the test? As someone who knows both optics and coding, do you have a feeling one way or the other? I also don't want to create extra work for you.

Comment (emolter) [reply jwst/ami/tests/test_matrix_dft.py:32]: Any future flexibility would be primarily for the benefit of your team, so if you're happy removing it then I'm happy - fewer code lines means less maintenance as far as I'm concerned

Comment (emolter) [reply jwst/ami/tests/test_matrix_dft.py:32]: Hi Rachel, looking at this again, it appears dropping the offset would be a larger change than I realized, because it is accessible from the top-level spec of the AmiAnalyzeStep.  That means it's "public API" as far as we're concerned, and users may theoretically change it themselves.

If you'd still prefer to see it removed, we can definitely do it, but in that case please open a separate JIRA ticket for discussion.

**Comment by emolter** [comment]
@anand0xff Thanks for looking into this.  I don't think the question is about the matrix DFT anymore - I am very glad you are happy with that portion.  Now, we are trying to figure out:

1. Whether or not the various options to `analyticnrm2.psf()` all work properly; in particular, whether all five options "circ", "circonly", "hex", "hexonly", and "fringeonly" produce accurate results, including for a non-zero input PSF offset.
2. Whether it is worth supporting a user-input PSF offset in the first place

File: jwst/ami/tests/test_matrix_dft.py (hunk 14 above)
Comment (rcooper295) [reply jwst/ami/tests/test_matrix_dft.py:32]: No, I think it should be fine to leave it in that case! Thanks for looking into it.

Comment (emolter) [reply jwst/ami/tests/test_matrix_dft.py:32]: ok thanks.  The small follow-up PR with the things we've talked about so far is at https://github.com/spacetelescope/jwst/pull/9441 let me know what you think.  Tyler had a question about the PSF function that I couldn't answer
